lib/rex/payloads/meterpreter/config.rb

```diff
@@ -0,0 +1,159 @@
+# -*- coding: binary -*-
+require 'msf/core/payload/uuid'
+require 'msf/core/payload/windows'
+require 'msf/core/reflective_dll_loader'
+require 'rex/parser/x509_certificate'
+
+class Rex::Payloads::Meterpreter::Config
+
+  include Msf::ReflectiveDLLLoader
+
+  UUID_SIZE = 64
+  URL_SIZE = 512
+  UA_SIZE = 256
+  PROXY_HOST_SIZE = 128
+  PROXY_USER_SIZE = 64
+  PROXY_PASS_SIZE = 64
+  CERT_HASH_SIZE = 20
+
+  def initialize(opts={})
+    @opts = opts
+    if opts[:ascii_str] && opts[:ascii_str] == true
+      @to_str = self.method(:to_ascii)
+    else
+      @to_str = self.method(:to_wchar_t)
+    end
+  end
+
+  def to_b
+    config_block
+  end
+
+private
+
+  def is_x86?
+    @opts[:arch] == ARCH_X86
+  end
+
+  def to_str(item, size)
+    @to_str.call(item, size)
+  end
+
+  def to_wchar_t(item, size)
+    to_ascii(item, size).unpack("C*").pack("v*")
+  end
+
+  def to_ascii(item, size)
+    item.to_s.ljust(size, "\x00")
+  end
+
+  def session_block(opts)
+    uuid = to_str(opts[:uuid].to_raw, UUID_SIZE)
+    if opts[:exitfunk]
+      exit_func = Msf::Payload::Windows.exit_types[opts[:exitfunk]]
+    else
+      exit_func = 0
+    end
+
+    session_data = [
+      0,                  # comms socket, patched in by the stager
+      exit_func,          # exit function identifer
+      opts[:expiration],  # Session expiry
+      uuid                # the UUID
+    ]
+
+    session_data.pack("VVVA*")
+  end
+
+  def transport_block(opts)
+    # Build the URL from the given parameters, and pad it out to the
+    # correct size
+    lhost = opts[:lhost]
+    if lhost && opts[:scheme].start_with?('http') && Rex::Socket.is_ipv6?(lhost)
+      lhost = "[#{lhost}]"
+    end
+
+    url = "#{opts[:scheme]}://#{lhost}:#{opts[:lport]}"
+    url << "?#{opts[:scope_id]}" if opts[:scope_id]
```

Comment:
I started using `?` as a cheap way of separating the `scope_id` out in the case of IPv6. In hindsight, it does seem rather silly to have them this way around.

BTW, may I request we have the non-stager related comments/feedback handled in #5300 please? :)